File: shared/Aegis/v2/aegis_py/aegis_py/server/interfaces/token_verifier.py
"""ITokenVerifier protocol — the interface all token verifier implementations must satisfy.

Aegis Python is server-only: Python services are always server-side, so there is no
client-side counterpart.
"""

from __future__ import annotations

from typing import Any, Protocol


class DecodedToken(Protocol):
    """Minimal interface for a decoded Firebase ID token.

    Attributes:
        uid: The user's unique identifier.
        email: The user's email address, if available.
        name: The user's display name, if available.
        picture: URL of the user's profile picture, if available.
    """

    @property
    def uid(self) -> str: ...

    @property
    def email(self) -> str | None: ...

    @property
    def name(self) -> str | None: ...

    @property
    def picture(self) -> str | None: ...


class ITokenVerifier(Protocol):
    """Interface for token verification implementations.

    Implementations must provide a single ``verify_token`` coroutine that accepts
    a raw bearer token string and returns decoded token data.
    """

    async def verify_token(self, token: str) -> Any:
        """Verify *token* and return the decoded payload.

        Args:
            token: The raw JWT bearer token string (without the ``Bearer `` prefix).

        Returns:
            A decoded token object (provider-specific type).

        Raises:
            ValueError: If the token is invalid, expired, or cannot be verified.
        """
        ...

File: shared/Aegis/v2/aegis_ts/src/server/interfaces/token-verifier.ts
/**
 * Decoded claims extracted from a verified authentication token.
 *
 * This type is intentionally provider-agnostic so that consumers never need to
 * import provider-specific SDKs.
 */
export interface DecodedToken {
  /** Unique identifier for the authenticated user. */
  readonly uid: string;
  /** Email address associated with the account, if available. */
  readonly email?: string;
  /** Display name of the user, if available. */
  readonly name?: string;
  /** URL of the user's profile picture, if available. */
  readonly picture?: string;
}

/**
 * Abstract interface for token verification operations.
 *
 * Implement this interface to swap the underlying authentication provider
 * without changing any business logic that depends on it.
 *
 * @example
 * ```typescript
 * // Implement for a new auth provider:
 * class Auth0TokenVerifier implements ITokenVerifier {
 *   async verifyToken(token: string): Promise<DecodedToken> { ... }
 * }
 * ```
 */
export interface ITokenVerifier {
  /**
   * Verifies an authentication token and returns the decoded claims.
   *
   * @param token - The raw authentication token string.
   * @returns Decoded token data including at minimum the user's uid.
   * @throws If the token is invalid, expired, or cannot be verified.
   */
  verifyToken(token: string): Promise<DecodedToken>;
}
